sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala

```diff
@@ -51,6 +51,11 @@ case class CostBasedJoinReorder(conf: CatalystConf) extends Rule[LogicalPlan] wi
 
   def reorder(plan: LogicalPlan, output: AttributeSet): LogicalPlan = {
     val (items, conditions) = extractInnerJoins(plan)
+    // Find the star schema joins. Currently, it returns the star join with the largest
+    // fact table. In the future, it can return more than one star join (e.g. F1-D1-D2
+    // and F2-D3-D4).
+    val starJoinPlans = StarSchemaDetection(conf).findStarJoins(items, conditions.toSeq)
```

Comment:
@wzhfy I’ve looked into moving the star reordering at the end of the optimization phase. Star reordering uses the existing ```ReorderJoin.createOrderedJoin``` method to construct the final plan once a star join is discovered. This method only handles specific types of plans, and doesn’t recognize the plan layout in the last phase of the Optimizer. Writing a new join reordering method for this purpose would not make too much sense since star joins are to be used by the existing planning strategies.

I suggest to keep the current logic and, next, I can look into integrating the star plans with your new DP planning. Once that’s tested, we can probably remove the star schema call from ```ReorderJoin``` planning rule. Please let me know what you think.